pkg/quota/admission/clusterresourceoverride/admission.go

```diff
@@ -0,0 +1,176 @@
+package clusterresourceoverride
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"reflect"
+
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+	configlatest "github.com/openshift/origin/pkg/cmd/server/api/latest"
+	"github.com/openshift/origin/pkg/project/cache"
+	"github.com/openshift/origin/pkg/quota/admission/clusterresourceoverride/api"
+	"github.com/openshift/origin/pkg/quota/admission/clusterresourceoverride/api/validation"
+	"k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/plugin/pkg/admission/limitranger"
+
+	"github.com/golang/glog"
+	"speter.net/go/exp/math/dec/inf"
+)
+
+const (
+	clusterResourceOverrideAnnotation = "quota.openshift.io/cluster-resource-override-enabled"
+	cpuBaseScaleFactor                = 1000.0 / (1024.0 * 1024.0 * 1024.0) // 1000 milliCores per 1GiB
+)
+
+func init() {
+	admission.RegisterPlugin(api.PluginName, func(client kclient.Interface, config io.Reader) (admission.Interface, error) {
+		return newClusterResourceOverride(client, config)
+	})
+}
+
+type internalConfig struct {
+	limitCPUToMemoryRatio     *inf.Dec
+	cpuRequestToLimitRatio    *inf.Dec
+	memoryRequestToLimitRatio *inf.Dec
+}
+type clusterResourceOverridePlugin struct {
+	*admission.Handler
+	internalConfig
+	Config       *api.ClusterResourceOverrideConfig
+	ProjectCache *cache.ProjectCache
+	LimitRanger  admission.Interface
+}
+
+var _ = oadmission.WantsProjectCache(&clusterResourceOverridePlugin{})
+var _ = oadmission.Validator(&clusterResourceOverridePlugin{})
+
+// newClusterResourceOverride returns an admission controller for containers that
+// configurably overrides container resource request/limits
+func newClusterResourceOverride(client kclient.Interface, config io.Reader) (admission.Interface, error) {
+	parsed, err := ReadConfig(config)
```

Comment:
On Tue, Feb 16, 2016 at 9:29 AM, Luke Meyer <lmeyer@redhat.com> wrote:
>
> The plugin init manager runs validation right after and kicks it out if it
> fails. Doing validation here seemed redundant, but I can if you like.
>

Actually since the original redundant Config has been removed from the
plugin object, doing validation here became essential.